palette/moreland/convert.go

```diff
@@ -0,0 +1,229 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package moreland
+
+import (
+	"image/color"
+	"math"
+)
+
+// rgb repesents a physically linear RGB color.
+type rgb struct {
+	R, G, B float64
+}
+
+// cieXYZ converts a linear RGB color to a CIE XYZ color.
+func (c rgb) cieXYZ() cieXYZ {
+	return cieXYZ{
+		X: 0.4124*c.R + 0.3576*c.G + 0.1805*c.B,
+		Y: 0.2126*c.R + 0.7152*c.G + 0.0722*c.B,
+		Z: 0.0193*c.R + 0.1192*c.G + 0.9505*c.B,
+	}
+}
+
+// sRGBA converts a linear RGB color to an sRGBA color, where alpha is the alpha
+// channel value (between 0 and 1).
+func (c rgb) sRGBA(alpha float64) sRGBA {
+	return sRGBA{
+		R: linearToS(c.R),
+		G: linearToS(c.G),
+		B: linearToS(c.B),
+		A: alpha,
+	}
+}
+
+// linearToS converts from a linear RGB component to an sRGB component.
+func linearToS(v float64) float64 {
+	if v > 0.0031308 {
+		return 1.055*math.Pow(v, 1/2.4) - 0.055
+	}
+	return 12.92 * v
+}
+
+// cieXYZ represents a color in CIE XYZ space.
+type cieXYZ struct {
+	X, Y, Z float64
+}
+
+// rgb converts a CIEXYZ color to physically linear RGB.
+func (c cieXYZ) rgb() rgb {
+	return rgb{
+		R: c.X*3.2406 + c.Y*-1.5372 + c.Z*-0.4986,
+		G: c.X*-0.9689 + c.Y*1.8758 + c.Z*0.0415,
+		B: c.X*0.0557 + c.Y*-0.204 + c.Z*1.057,
+	}
+}
+
+// cieLAB converts an CIE XYZ color to a CIE LAB color.
+func (c cieXYZ) cieLAB() cieLAB {
+	tempX := xyz2labHelper(c.X / 0.9505)
+	tempY := xyz2labHelper(c.Y)
+	tempZ := xyz2labHelper(c.Z / 1.089)
+	return cieLAB{
+		L: (116.0 * tempY) - 16.0,
+		A: 500.0 * (tempX - tempY),
+		B: 200 * (tempY - tempZ),
+	}
+}
+
+// xyz2labHelper is an intermediate step in converting from CIE XYZ to CIE LAB.
+func xyz2labHelper(v float64) float64 {
```

Comment:
This name is still unsatisfying, but it seems that Mooreland is actually using the correct name, `f`. So define a var func `f` in cieLAB above and use that.